### PR TITLE
refactor(tree-select): rewrite with hook

### PR DESCRIPTION
**Pull request by hengkx**
### 🤔 This is a ...

- [ ] New feature
- [ ] Bug fix
- [ ] Site / document update
- [ ] Component style update
- [ ] TypeScript definition update
- [x] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Other (about what?)

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |  refactor with hook         |
| 🇨🇳 Chinese |  用 react hook 重构         |

### ☑️ Self Check before Merge

⚠️ Please check all items below before review. ⚠️ 

- [x] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [x] TypeScript definition is updated/provided or not needed
- [x] Changelog is provided or not needed

### CODE REVIEW

File: components/tree-select/index.tsx [hunk 1]
```diff
@@ -35,162 +35,137 @@ export interface TreeSelectProps<T>
...
 
+interface CompoundedComponent<T>
+  extends React.ForwardRefExoticComponent<TreeSelectProps<T> & React.RefAttributes<HTMLElement>> {
+  TreeNode: typeof TreeNode;
+  SHOW_ALL: typeof SHOW_ALL;
+  SHOW_PARENT: typeof SHOW_PARENT;
+  SHOW_CHILD: typeof SHOW_CHILD;
+}
+
+const TreeSelect = React.forwardRef(InternalTreeSelect) as CompoundedComponent<unknown>;
```
Comment (hengkx) [inline components/tree-select/index.tsx:158]: @afc163  这一块TS有问题 请教下 可以怎么写？

Comment (zombieJ) [reply components/tree-select/index.tsx:158]: Select 和 TreeSelect 重写还是用 class component 是 FC 用 forwardRef 不支持 generic……

Comment (zombieJ) [reply components/tree-select/index.tsx:158]: ref https://github.com/react-component/select/blob/12a0859a036781666151f042afa6f5e2b6e44851/src/Select.tsx#L70

Comment (hengkx) [reply components/tree-select/index.tsx:158]: 怪不得 找了半天 没找到合适的方式 ts后续会有说支持吗？

Comment (zombieJ) [reply components/tree-select/index.tsx:158]: 暂时没看到相关消息。从 React 官方看，FC 和 CC 是并存关系，FC 不行就换 CC 所以也不是特别大的问题。

**Pull request closed without merging**